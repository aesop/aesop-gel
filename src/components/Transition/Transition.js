import React, { isValidElement } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import get from 'lodash/get';
import { CSSTransition } from 'react-transition-group';
import useHasMounted from '~/customHooks/useHasMounted';
import data from './Transition.data';

const Transition = ({
  children,
  hasCSSTransitionMountOnEnter,
  hasCSSTransitionUnmountOnExit,
  isActiveOnMount,
  isActive,
  type,
}) => {
  const hasMounted = useHasMounted();

  if (!isValidElement(children)) {
    return null;
  }

  const isVisible = isActiveOnMount ? !!hasMounted : !!isActive;

  return (
    <CSSTransition
      classNames={get(data[type], 'classNames', '')}
      in={isVisible}
      mountOnEnter={!!hasCSSTransitionMountOnEnter}
      timeout={get(data[type], 'timeout', 300)}
      unmountOnExit={!!hasCSSTransitionUnmountOnExit}
    >
      {React.cloneElement(children, {
        className: cx(
          get(children, 'props.className', ''),
          isVisible ? '' : get(data[type], 'classNames.enter', ''),
        ),
      })}
    </CSSTransition>
  );
};

Transition.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element]).isRequired,
  hasCSSTransitionMountOnEnter: PropTypes.bool,
  hasCSSTransitionUnmountOnExit: PropTypes.bool,
  isActive: PropTypes.bool,
  isActiveOnMount: PropTypes.bool,
  type: PropTypes.oneOf([
    'fade',
    'shiftInDown',
    'shiftInLeft',
    'slideDown',
    'slideRight',
    'slowFade',
    'zoom',
  ]).isRequired,
};

Transition.defaultProps = {
  children: undefined,
  hasCSSTransitionMountOnEnter: undefined,
  hasCSSTransitionUnmountOnExit: undefined,
  isActive: false,
  isActiveOnMount: false,
  type: undefined,
};

export default Transition;
