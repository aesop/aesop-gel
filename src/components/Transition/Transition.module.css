@import 'styles/variables/index';

.fadeEnter {
  opacity: 0;
}

.fadeEnterActive {
  opacity: 1;
  transition: opacity 500ms var(--easing-ease-out-cubic);
}

.fadeEnterDone {
  opacity: 1;
}

.fadeExit {
  opacity: 1;
}

.fadeExitActive {
  opacity: 0;
  transition: opacity 500ms var(--easing-ease-out-cubic);
}

.fadeExitDone {
  opacity: 0;
}

.shiftInDownEnter {
  opacity: 0;
  transform: translate3d(0, -20px, 0);
}

.shiftInDownEnterActive {
  opacity: 1;
  transform: translate3d(0, 0, 0);
  transition:
    transform 1500ms var(--easing-ease-out-cubic),
    opacity 1500ms var(--easing-ease-in-cubic);
}

.shiftInDownEnterDone {
  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInDownExit {
  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInDownExitActive {
  opacity: 0;
  transform: translate3d(0, -20px, 0);
  transition:
    transform 1500ms var(--easing-ease-out-cubic),
    opacity 1500ms var(--easing-ease-in-cubic);
}

.shiftInDownExitDone {
  opacity: 0;
  transform: translate3d(0, -20px, 0);
}

.shiftInLeftEnter {
  opacity: 0;
  transform: translate3d(20px, 0, 0);
}

.shiftInLeftEnterActive {
  opacity: 1;
  transform: translate3d(0, 0, 0);
  transition:
    transform 1500ms var(--easing-ease-out-cubic),
    opacity 1000ms var(--easing-ease-in-cubic);
}

.shiftInLeftEnterDone {
  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInLeftExit {
  opacity: 1;
  transform: translate3d(0, 0, 0);
}

.shiftInLeftExitActive {
  opacity: 0;
  transform: translate3d(20px, 0, 0);
  transition:
    transform 1500ms var(--easing-ease-out-cubic),
    opacity 1000ms var(--easing-ease-in-cubic);
}

.shiftInLeftExitDone {
  opacity: 0;
  transform: translate3d(20px, 0, 0);
}

.slideDownEnter {
  transform: translate3d(0, -100%, 0);
  visibility: hidden;
}

.slideDownEnterActive {
  transform: translate3d(0, 0, 0);
  transition:
    transform 500ms var(--easing-ease-out-cubic),
    visibility 500ms var(--easing-ease-out-cubic);
  visibility: visible;
}

.slideDownEnterDone {
  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideDownExit {
  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideDownExitActive {
  transform: translate3d(0, -100%, 0);
  transition:
    transform 500ms var(--easing-ease-out-cubic),
    visibility 500ms var(--easing-ease-out-cubic);
  visibility: hidden;
}

.slideDownExitDone {
  transform: translate3d(0, -100%, 0);
  visibility: hidden;
}

.slideRightEnter {
  transform: translate3d(-100%, 0, 0);
  visibility: hidden;
}

.slideRightEnterActive {
  transform: translate3d(0, 0, 0);
  transition:
    transform 500ms var(--easing-ease-out-cubic),
    visibility 500ms var(--easing-ease-out-cubic);
  visibility: visible;
}

.slideRightEnterDone {
  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideRightExit {
  transform: translate3d(0, 0, 0);
  visibility: visible;
}

.slideRightExitActive {
  transform: translate3d(-100%, 0, 0);
  transition:
    transform 500ms var(--easing-ease-out-cubic),
    visibility 500ms var(--easing-ease-out-cubic);
  visibility: hidden;
}

.slideRightExitDone {
  transform: translate3d(-100%, 0, 0);
  visibility: hidden;
}

.slowFadeEnter {
  opacity: 0;
}

.slowFadeEnterActive {
  opacity: 1;
  transition: opacity 1500ms var(--easing-ease-out-cubic);
}

.slowFadeEnterDone {
  opacity: 1;
}

.slowFadeExit {
  opacity: 1;
}

.slowFadeExitActive {
  opacity: 0;
  transition: opacity 1500ms var(--easing-ease-out-cubic);
}

.slowFadeExitDone {
  opacity: 0;
}

.zoomEnter {
  opacity: 0;
  transform: scale3d(0.9, 0.9, 0.9);
}

.zoomEnterActive {
  opacity: 1;
  transform: scale3d(1, 1, 1);
  transition:
    transform 500ms var(--easing-ease-out-cubic),
    opacity 500ms var(--easing-ease-out-cubic);
}

.zoomEnterDone {
  opacity: 1;
  transform: scale3d(1, 1, 1);
}

.zoomExit {
  opacity: 1;
  transform: scale3d(1, 1, 1);
}

.zoomExitActive {
  opacity: 0;
  transform: scale3d(0.9, 0.9, 0.9);
  transition:
    transform 500ms var(--easing-ease-out-cubic),
    opacity 500ms var(--easing-ease-out-cubic);
}

.zoomExitDone {
  opacity: 0;
  transform: scale3d(0.9, 0.9, 0.9);
}
