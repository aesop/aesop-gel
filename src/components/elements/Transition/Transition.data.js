import transitions from './css/transitions.module.css';

const fade = {
  classNames: {
    enter: transitions.fadeEnter,
    enterActive: transitions.fadeEnterActive,
    exit: transitions.fadeExit,
    exitActive: transitions.fadeExitActive,
  },
  timeout: 500,
};

const slowFade = {
  classNames: {
    enter: transitions.slowFadeEnter,
    enterActive: transitions.slowFadeActive,
    exit: transitions.slowFadeExit,
    exitActive: transitions.slowFadeExitActive,
  },
  timeout: 1500,
};

const shiftInLeft = {
  classNames: {
    enter: transitions.shiftInLeftEnter,
    enterActive: transitions.shiftInLeftActive,
    exit: transitions.shiftInLeftExit,
    exitActive: transitions.shiftInLeftExitActive,
  },
  timeout: 1500,
};

const shiftInDown = {
  classNames: {
    enter: transitions.shiftInDownEnter,
    enterActive: transitions.shiftInDownActive,
    exit: transitions.shiftInDownExit,
    exitActive: transitions.shiftInDownExitActive,
  },
  timeout: 1500,
};

const slideDown = {
  classNames: {
    enter: transitions.slideDownEnter,
    enterActive: transitions.slideDownEnterActive,
    exit: transitions.slideDownExit,
    exitActive: transitions.slideDownExitActive,
  },
  timeout: 500,
};

const slideRight = {
  classNames: {
    enter: transitions.slideRightEnter,
    enterActive: transitions.slideRightEnterActive,
    exit: transitions.slideRightExit,
    exitActive: transitions.slideRightExitActive,
  },
  timeout: 500,
};

const zoom = {
  classNames: {
    enter: transitions.zoomEnter,
    enterActive: transitions.zoomEnterActive,
    exit: transitions.zoomExit,
    exitActive: transitions.zoomExitActive,
  },
  timeout: 500,
};

export default {
  fade,
  shiftInDown,
  shiftInLeft,
  slideDown,
  slideRight,
  slowFade,
  zoom,
};
