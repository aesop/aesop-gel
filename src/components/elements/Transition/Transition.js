import React from 'react';
import { CSSTransition } from 'react-transition-group';
import useHasMounted from '~/hooks/useHasMounted';
import data from './Transition.data';
import Props from './Transition.types';

const Transition: React.FunctionComponent<Props> = ({
  active,
  activeOnMount,
  children,
  mountOnEnter,
  type,
  unmountOnExit,
}) => {
  const hasMounted = useHasMounted();

  return (
    <CSSTransition
      classNames={data[type].classNames}
      in={activeOnMount ? hasMounted : active}
      mountOnEnter={mountOnEnter}
      timeout={data[type].timeout}
      unmountOnExit={unmountOnExit}
    >
      {children}
    </CSSTransition>
  );
};

Transition.defaultProps = {
  active: true,
  activeOnMount: false,
  mountOnEnter: false,
  unmountOnExit: false,
};

export default Transition;
