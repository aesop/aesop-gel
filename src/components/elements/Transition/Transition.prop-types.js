import PropTypes from 'prop-types';
import { getPropTypeOptionValues } from '~/utils/propTypes';

export const TRANSITION_PROP_TYPE_OPTIONS = {
  FADE: 'fade',
  SHIFT_IN_DOWN: 'shiftInDown',
  SHIFT_IN_LEFT: 'shiftInLeft',
  SLIDE_DOWN: 'slideDown',
  SLIDE_RIGHT: 'slideRight',
  SLOW_FADE: 'slowFade',
  ZOOM: 'zoom',
};

export const PROP_TYPES = {
  active: PropTypes.bool,
  activeOnMount: PropTypes.bool,
  children: PropTypes.node.isRequired,
  mountOnEnter: PropTypes.bool,
  type: PropTypes.oneOf(getPropTypeOptionValues(TRANSITION_PROP_TYPE_OPTIONS))
    .isRequired,
  unmountOnExit: PropTypes.bool,
};

export default PROP_TYPES;
