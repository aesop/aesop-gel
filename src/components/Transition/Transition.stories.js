import React from 'react';
import { storiesOf } from '@storybook/react';
import { select } from '@storybook/addon-knobs';
import { State } from '~.storybook/State';
import Button from '~/components/Button';
import DefinitionList from '~/components/DefinitionList';
import Transition from './Transition';

storiesOf('Transition', module).add('Base component', () => (
  <State initialState={{ isActive: false }}>
    {(state, setState) => (
      <>
        <Button
          onClick={() => {
            setState(state => ({ ...state, isActive: !state.isActive }));
          }}
          title="Toggle Transition"
        >
          Toggle Transition
        </Button>
        <br />
        <br />
        <Transition
          isActive={state.isActive}
          type={select(
            'type',
            [
              'fade',
              'shiftInDown',
              'shiftInLeft',
              'slideDown',
              'slideRight',
              'slowFade',
              'zoom',
            ],
            'shiftInDown',
          )}
        >
          <DefinitionList
            items={[
              {
                term: 'Aroma',
                description: 'Fresh, woody, citrus',
              },
              {
                term: 'Usage',
                description:
                  'Spray  two to three pumps throughout the immediate space and refresh as needed; the aroma will last for several hours.',
              },
            ]}
          />
        </Transition>
      </>
    )}
  </State>
));
