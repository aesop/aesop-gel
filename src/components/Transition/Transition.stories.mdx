import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { withKnobs } from '@storybook/addon-knobs';
import * as knobs from '@storybook/addon-knobs';
import { State } from '~.storybook/State';
import Button from '~/components/Button';
import DefinitionList from '~/components/DefinitionList';
import DefinitionListFixture from '~/components/DefinitionList/DefinitionList.fixture';
import Transition from './Transition';

<Meta
  title="Components/Transition"
  component={Transition}
  decorators={[withKnobs]}
/>

# Transition

<Preview>
  <Story name="Base component">
    <State initialState={{ isActive: false }}>
      {(state, setState) => {
        return (
          <React.Fragment>
            <Button
              onClick={() => {
                setState(state => ({ ...state, isActive: !state.isActive }));
              }}
              title="Toggle Transition"
            >
              Toggle Transition
            </Button>
            <Transition
              isActive={state.isActive}
              type={knobs.select(
                'type',
                [
                  'fade',
                  'shiftInDown',
                  'shiftInLeft',
                  'slideDown',
                  'slideRight',
                  'slowFade',
                  'zoom',
                ],
                'shiftInDown',
              )}
            >
              <DefinitionList
                items={DefinitionListFixture.items}
              />
            </Transition>
          </React.Fragment>
        )
      }}
    </State>
  </Story>
</Preview>

## API

<Props of={Transition} />
