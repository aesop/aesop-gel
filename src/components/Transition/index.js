/* istanbul ignore file */
import Transition from './Transition';

export default Transition;
