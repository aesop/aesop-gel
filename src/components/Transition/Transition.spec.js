import React from 'react';
import renderer from 'react-test-renderer';
import Transition from './Transition';

describe('<Transition />', () => {
  it('should be defined', () => {
    expect(Transition).toBeDefined();
  });

  it('renders base component correctly', () => {
    const tree = renderer
      .create(
        <Transition isActive={true} type="fade">
          <div>Content</div>
        </Transition>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
