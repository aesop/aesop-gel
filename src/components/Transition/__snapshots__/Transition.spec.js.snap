// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Transition /> renders base component correctly 1`] = `
Array [
  <div
    className=""
  >
    Content
  </div>,
  <div>
    Content
  </div>,
]
`;
